### PR TITLE
[16.0][ADD] sale_stock_warehouse_multicompany 

### PR DESCRIPTION
**Pull request by santostelmo**
Allow multiple companies to sell the stock of a shared warehouse.

The sales order will belong to the company making the sale. As such, the taxes and related invoices will also belong to that selling company. The generated stock moves will belong to the company owning the stock and defined on the warehouse.

This approach also allows to use any route (like cross-docking or drop-shipment) on the sales order line.

PR superseded https://github.com/OCA/multi-company/pull/573

### CODE REVIEW

File: sale_stock_warehouse_multicompany/hooks.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    model = env["stock.warehouse"]
+    table_name = model._fields["company_ids"].relation
+    column1 = model._fields["company_ids"].column1
+    column2 = model._fields["company_ids"].column2
+    SQL = """
+        INSERT INTO {}
+        ({}, {})
+        SELECT id, company_id FROM {} WHERE company_id IS NOT NULL
+        ON CONFLICT DO NOTHING
+    """.format(
```
Comment (simahawk) [inline sale_stock_warehouse_multicompany/hooks.py]: do not use format

File: sale_stock_warehouse_multicompany/readme/CONTRIBUTORS.rst [hunk 2]
```diff
@@ -0,0 +1,18 @@
+* Odoo S.A. (original module `inter_company_rules`)
```
Comment (simahawk) [inline sale_stock_warehouse_multicompany/readme/CONTRIBUTORS.rst]: this list seems wrong

Comment (santostelmo) [reply sale_stock_warehouse_multicompany/readme/CONTRIBUTORS.rst]: Yes is wrong, still working on the PR

File: sale_stock_warehouse_multicompany/hooks.py (hunk 1 above)
Comment (santostelmo) [reply sale_stock_warehouse_multicompany/hooks.py]: That was taken from the hooks example present in the base module. I'll change it anyway

Comment (santostelmo) [reply sale_stock_warehouse_multicompany/hooks.py]: what do you suggest to use ?

Comment (simahawk) [reply sale_stock_warehouse_multicompany/hooks.py]: https://www.psycopg.org/psycopg3/docs/basic/params.html

File: sale_stock_warehouse_multicompany/models/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+from odoo import fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    warehouse_id = fields.Many2one(
+        "stock.warehouse", 
+        domain="[('company_ids', 'in', company_id)]",
+        check_company = False
```
Comment (jbaudoux) [inline sale_stock_warehouse_multicompany/models/sale_order.py]: You sure you need this check_company=False when you set a domain?
See https://github.com/odoo/odoo/blob/16.0/odoo/fields.py#L2834
You have not done that on the SO line

File: sale_stock_warehouse_multicompany/readme/CONTRIBUTORS.rst [hunk 4]
```diff
@@ -0,0 +1,2 @@
+  * Cyril Manuel Jeanneret <cyril.jeanneret@camptocamp.com>
+  * Telmo Santos <telmo.santos@camptocamp.com>
```
Comment (jbaudoux) [inline sale_stock_warehouse_multicompany/readme/CONTRIBUTORS.rst]:
```suggestion
  * Telmo Santos <telmo.santos@camptocamp.com>
  * Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
```

**Review by jbaudoux — COMMENTED** [review]
LG. You also need to adapt SO line `_prepare_procurement_values` to set company of the warehouse and not of the SO

File: sale_stock_warehouse_multicompany/models/sale_order.py (hunk 3 above)
Comment (santostelmo) [reply sale_stock_warehouse_multicompany/models/sale_order.py]: Checked together. We need that

**Comment by santostelmo** [comment]
> LG. You also need to adapt SO line `_prepare_procurement_values` to set company of the warehouse and not of the SO

Done. Indeed we need to create procurements in the warehouse company and not sale order company.

File: sale_stock_warehouse_multicompany/models/sale_order_line.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    route_id = fields.Many2one(
+        "stock.route", domain="[('company_ids', 'in', company_id)]"
```
Comment (jbaudoux) [inline sale_stock_warehouse_multicompany/models/sale_order_line.py]:
```suggestion
        "stock.route", domain="[('company_ids', 'in', company_id)]",
        check_company=False,
```

**Review by jbaudoux — CHANGES_REQUESTED** [review]

File: sale_stock_warehouse_multicompany/models/sale_order_line.py (hunk 5 above)
Comment (santostelmo) [reply sale_stock_warehouse_multicompany/models/sale_order_line.py]: Change done

File: sale_stock_warehouse_multicompany/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+{
+    "name": "sale stock warehouse multicompany",
+    "summary": "Sale stock warehouse multicompany",
```
Comment (jbaudoux) [inline sale_stock_warehouse_multicompany/__manifest__.py]:
```suggestion
    "summary": "Allow multiple companies to sell the stock of a shared warehouse",
```

File: sale_stock_warehouse_multicompany/models/sale_order_line.py (hunk 5 above)
Comment (jbaudoux) [reply sale_stock_warehouse_multicompany/models/sale_order_line.py]: I don't see it

Comment (santostelmo) [reply sale_stock_warehouse_multicompany/models/sale_order_line.py]: Now it should be there

**Review by jbaudoux — APPROVED** [review]

**Review by sebalix — APPROVED** [review]

**Comment by sebalix** [comment]
Please rename the module `sale_stock_warehouse_multicompany` in the PR title

**Review by cyrilmanuel — APPROVED** [review]

**Comment by cyrilmanuel** [comment]
hiii @pedrobaeza can you merge this PR please ?

**Pull request merged**